sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorDispatcher.java

```diff
@@ -149,9 +150,14 @@ private void signalWorkQueue() throws IOException {
         } catch (ClosedChannelException ignorePipeClosedDuringReactorShutdown) {
             if (!isClosed.get()) {
                 logger.warning("signalWorkQueue failed before reactor closed.", ignorePipeClosedDuringReactorShutdown);
-                shutdownSignal.emitError(new RuntimeException(String.format(
-                    "connectionId[%s] IO Sink was interrupted before reactor closed.", connectionId),
-                    ignorePipeClosedDuringReactorShutdown), Sinks.EmitFailureHandler.FAIL_FAST);
+                if (!isIOPipeInterrupted.getAndSet(true)) {
+                    shutdownSignal.emitError(new RuntimeException(String.format(
+                        "connectionId[%s] IO Sink was interrupted before reactor closed.", connectionId),
+                        ignorePipeClosedDuringReactorShutdown), Sinks.EmitFailureHandler.FAIL_FAST);
+                } else {
+                    throw logger.logExceptionAsWarning(new RejectedExecutionException(String.format(
```

Comment:
Do we need to rethrow this exception? I was wondering if this would add additional noise to the stack. Is logging it enough since a shutdown signal was already emitted?

Reply:
What I observed is - while a shutdown signal is emitted, there are Code-Paths that call invoke concurrently and result CCE. Code-Paths such as adding credit, session sub-graph cleanup, etc., and CCE are silently dropped in those cases, resulting in the "Sinks" those Code-Paths are supposed to notify never triggered. We indeed have the timeout for those Code-Paths, but the timeout is not completing the wrapped "Sinks", which hangs downstream. This is what I observed.

Reply:
I'm continuing investigation in this space will document details/push change

Reply:
Another reason I thought throwing may make sense was - aligning with the current behavior. I.e., today by design, any thread that comes when the pipe is closed gets "RejectedException".

The following log from a Cx is an example -

An interrupted thread T1 results pipe-closure and logged below - 

>  {"timestamp":"1645535822119","logger":"com.azure.core.amqp.implementation.ReactorDispatcher","thread":"parallel-2","level":"WARN","message":"connectionId[MF_bee78a_1645405261748] signalWorkQueue failed before reactor closed."}


A later thread T2 receives "RejectedException" because the pipe was closed

> {"timestamp":"1645535882169","logger":"com.azure.core.amqp.implementation.ReactorDispatcher","thread":"boundedElastic-36","level":"WARN","message":"ReactorDispatcher instance is closed. Should not continue dispatching work to this reactor."}

So, this PR is trying to adhere to this today's design, i.e., the Threads that couldn't use closed-pipe (because an interrupt in a "concurrent" thread closed the pipe) get "RejectedException".